File: src/types/graph.ts
export type NodeType = 'intersection' | 'shelter' | 'hospital' | 'depot' | 'residential' | 'commercial';

export interface GraphNode {
  id: string;
  name: string;
  lat: number;
  lng: number;
  type: NodeType;
  elevation: number; // in meters
  population: number; // current population at node
  evacuatedCount?: number;
  capacity?: number; // for shelters/hospitals
  currentOccupancy?: number;
  hazardRisk: number; // 0 (safe) to 1 (lethal)
  zoneId?: string; // for phased evacuation
  color?: string; // for vertex coloring
  isDominatingHub?: boolean; // for dominating set
  isDistressActive?: boolean;
  distressPriority?: 'P1' | 'P2' | 'P3';
}

export type RoadType = 'highway' | 'arterial' | 'local' | 'bridge' | 'tunnel' | 'mountain_pass';

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  distance: number; // in km
  baseSpeed: number; // in km/h
  capacity: number; // vehicle/person flow capacity per hour
  currentFlow: number; // for max-flow calculations
  roadType: RoadType;
  lanes: number;
  elevationSlope: number; // gradient percentage
  hazardRisk: number; // 0.0 to 1.0
  predictedRisk: number; // AI predicted risk in next dt
  isBlocked: boolean;
  blockageReason?: string;
  isBridge?: boolean; // detected by Tarjan's algorithm
  isMstEdge?: boolean; // part of MST
  isMinCut?: boolean; // bottleneck edge in max-flow
  inActivePath?: boolean; // part of computed escape path
  trafficDensity: number; // 0.0 (empty) to 1.0 (gridlock)
}

export interface GraphData {
  nodes: Record<string, GraphNode>;
  edges: Record<string, GraphEdge>;
  adjacencyList: Record<string, string[]>; // node -> edge IDs
}

export type DisasterType = 'flood' | 'earthquake' | 'wildfire' | 'landslide' | 'cyclone';

export interface HazardZone {
  id: string;
  type: DisasterType;
  centerLat: number;
  centerLng: number;
  radiusKm: number;
  intensity: number; // 0 to 1
  expansionRateKmH: number; // growth per hour
  directionDeg?: number; // propagation heading degrees
  windAngleDeg?: number; // for wildfire / cyclone (0 = East, 90 = North)
  windSpeedKmH?: number;
  waterLevelMeters?: number; // for floods
  epicenterDepthKm?: number; // for earthquakes
}

export interface RouteResult {
  pathNodeIds: string[];
  pathEdgeIds: string[];
  totalDistanceKm: number;
  estimatedTimeMin: number;
  safetyScore: number; // 0 to 100
  hazardExposureRisk: number; // aggregate risk
  isPassable: boolean;
  bottlenecks: string[];
  steps: {
    instruction: string;
    distanceKm: number;
    roadName: string;
    hazardWarning?: string;
  }[];
}

export interface AlgorithmStep {
  stepIndex: number;
  description: string;
  highlightedNodeIds: string[];
  highlightedEdgeIds: string[];
  visitedNodeIds?: string[];
  currentDistanceMap?: Record<string, number>;
  metrics?: Record<string, string | number>;
}

export interface AlgorithmExecutionResult {
  algorithmName: string;
  executionTimeMs: number;
  summary: string;
  steps: AlgorithmStep[];
  nodeAttributes?: Record<string, Partial<GraphNode>>;
  edgeAttributes?: Record<string, Partial<GraphEdge>>;
  customMetrics?: Record<string, string | number>;
  additionalPaths?: { id: string; label: string; edgeIds: string[]; color: string }[];
}

File: backend/models/graph.py
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field

NodeType = Literal['intersection', 'shelter', 'hospital', 'depot', 'residential', 'commercial']
RoadType = Literal['highway', 'arterial', 'local', 'bridge', 'tunnel', 'mountain_pass']
DisasterType = Literal['flood', 'earthquake', 'wildfire', 'landslide', 'cyclone']

class GraphNode(BaseModel):
    id: str
    name: str
    lat: float
    lng: float
    type: NodeType
    elevation: float = 0.0
    population: int = 0
    evacuatedCount: Optional[int] = 0
    capacity: Optional[int] = None
    currentOccupancy: Optional[int] = 0
    hazardRisk: float = 0.0
    zoneId: Optional[str] = None
    color: Optional[str] = None
    isDominatingHub: Optional[bool] = False
    isDistressActive: Optional[bool] = False
    distressPriority: Optional[Literal['P1', 'P2', 'P3']] = None

class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    distance: float
    baseSpeed: float = 50.0
    capacity: float = 1000.0
    currentFlow: float = 0.0
    roadType: RoadType = 'local'
    lanes: int = 2
    elevationSlope: float = 0.0
    hazardRisk: float = 0.0
    predictedRisk: float = 0.0
    isBlocked: bool = False
    blockageReason: Optional[str] = None
    isBridge: Optional[bool] = False
    isMstEdge: Optional[bool] = False
    isMinCut: Optional[bool] = False
    inActivePath: Optional[bool] = False
    trafficDensity: float = 0.0

class GraphData(BaseModel):
    nodes: Dict[str, GraphNode]
    edges: Dict[str, GraphEdge]
    adjacencyList: Dict[str, List[str]]

class HazardZone(BaseModel):
    id: str
    type: DisasterType
    centerLat: float
    centerLng: float
    radiusKm: float
    intensity: float = 0.5
    expansionRateKmH: float = 0.5
    directionDeg: Optional[float] = 0.0
    windAngleDeg: Optional[float] = None
    windSpeedKmH: Optional[float] = None
    waterLevelMeters: Optional[float] = None
    epicenterDepthKm: Optional[float] = None
